### DISCUSSION TITLE
Analog to request.get_data() from flask?

### DISCUSSION THREAD

**Discussion by rlewkowicz** [fastapi/fastapi#8112] [Questions]
I pickle an image and fire off using python requests:

```
data = pickle.dumps(img)
r = requests.post(url,data=data)
```
Then in flask:
```
image = pickle.loads(request.get_data())
```

Super simple. In flask, data is where it puts something that it doesn't know how to handle. From the docs:
```
Contains the incoming request data as string in case it came with a mimetype Werkzeug does not handle.
```
This is good because I don't know what a serialized object would be as far as a mime type. Two, I don't know what the requests library is doing when I do requests.post under the hood. So it's nice to just throw crap at flask and not worry about what it is. 

Then your docs have accessing request.client.whatever and thats cool, but then I'm trying to get the body (I think) of this starlette object and they're doing async things and I don't know what this is in the context of fast api.

```
https://www.starlette.io/requests/#body
async def app(scope, receive, send):
    assert scope['type'] == 'http'
    request = Request(scope, receive)
    body = b''
    async for chunk in request.stream():
        body += chunk
    response = Response(body, media_type='text/plain')
    await response(scope, receive, send)
```
I would like to just throw whatever at fast api and grab it on the other side. Any tips on how to do that?

**Comment by euri10** [comment]
I think, if I get correctly what you want because it;s not very clear to me what get_data() does nor where you'd want to use it, that you can use something like `body = await request.body()`

the below dumps 2 files, pickle them, send them in a post endpoint that loads the body and guess the mime and returns it : that's what I understood you want to achieve ...

```
import pickle

import magic
import pytest
from async_asgi_testclient import TestClient
from fastapi import FastAPI
from starlette.requests import Request


# r = requests.post(url,data=data)

app = FastAPI()


@app.post("/image")
async def image_create(request: Request):
    body = await request.body()
    l = pickle.loads(body)
    mime = magic.from_buffer(l)
    return {"mime": mime}


img_data = [("raw.gif", "GIF image data"), ("raw.jpeg", "JPEG image data")]


@pytest.mark.parametrize("img, expected_mime", img_data)
@pytest.mark.asyncio
async def test_mime(img, expected_mime):
    with open(img, "rb") as im:
        r = im.read()
        p = pickle.dumps(r)
    async with TestClient(app) as client:
        response = await client.post("/image", data=p)
        assert response.status_code == 200
        assert expected_mime in response.json().get("mime")

```

**Comment by rlewkowicz** [comment]
Thanks @euri10. This most likely would have worked but I switched over to web sockets. I know it's bad form, but I'm going to sneak another question to you here if you're willing to answer it. 

How do you share an object, a queue or a variable between definitions? How do I have a global object in this fastapi structure. I'm not worried about it being thread safe and I'm running into issues trying to just do a global variable.

**Comment by euri10** [comment]
what issues ? 

random ideas just in case, not being sure exactly what you meant by global.

- use app.state should #502 be implemented or use @dmontagu way of doing the same without the PR implemented (https://github.com/tiangolo/fastapi/issues/508#issuecomment-537841234)

- should you need something more sophisticated you can take a look at contextvars blackmagic https://github.com/tiangolo/fastapi/issues/397

**Comment by dmontagu** [comment]
@rlewkowicz why can’t you just use a regular global variable if that’s what you want?

**Comment by rlewkowicz** [comment]
@dmontagu First I have to clarify I'm not the best developer (this is just a hobby project) and may be misunderstanding things. 

This is essentially a hacked up version of https://github.com/ultralytics/yolov3 

In this, I have "plots" Its just a couple of tuples and a confidence level. If plots is defined in the method, works great. If plots is defined in a global scope, it seems to just have a variety of issues. One of my problems was on me, but now it just seems to hang. If I define plots in a global scope and then access it from input and detect it will just hang

```
@app.websocket("/detect")
async def websocket_endpoint(websocket: WebSocket):
    try:
        await websocket.accept()
        while True:
            t0 = time.time()
            rec = await websocket.receive_bytes()
            image = msgpack.unpackb(rec)
            image = cv2.cvtColor(image, cv2.COLOR_BGR2RGB)
            img, im0 = LoadImages(image=image, img_size=img_size, half=False)

            img = torch.from_numpy(img).to(device)
            if img.ndimension() == 3:
                img = img.unsqueeze(0)
            pred, _ = model(img)
            plots = []
            for i, det in enumerate(non_max_suppression(pred, 0.1, 0.5)): 
                s=''
                s += '%gx%g ' % img.shape[2:]  # print string
                print(s)
                if det is not None and len(det):
                    # Rescale boxes from img_size to im0 size
                    det[:, :4] = scale_coords(img.shape[2:], det[:, :4], im0.shape).round()

                    # Print results
                    for c in det[:, -1].unique():
                        n = (det[:, -1] == c).sum()  # detections per class
                        s += '%g %ss, ' % (n, classes[int(c)])  # add to string
                    for *xyxy, conf, _, cls in det:
                        label = '%.2f' % (conf)
                        plots.append([return_plot(xyxy, im0),float(label)])
            try:
                print()
                tl = round(0.002 * (image.shape[0] + image.shape[1]) / 2) + 1  # line thickness
                color = [random.randint(0, 255) for _ in range(3)]
                cv2.rectangle(image, tuple(plots[0][0][0]), tuple(plots[0][0][1]), color, thickness=tl)
                cv2.imshow('en', image)
                cv2.waitKey(1)
            except:
                pass

            t = time.time()
            print('Done. (%.3fs)' % (time.time() - t0))
    except:
        pass
            

@app.websocket("/input")
async def input_endpoint(websocket: WebSocket):
    await websocket.accept()
    while True:
        rec = await websocket.receive_bytes()
        x=plots[0][0][0][0] - 208
        y=plots[0][0][0][1] - 208
        ser.write((str(x)+'h'+str(y) + '\n').encode())
```

**Comment by dmontagu** [comment]
@rlewkowicz I think what's happening is the following.

Since you assign to `plots` in the function, python treats it as locally scoped:
```python
            plots = []
```

Try the following:
```python
plots = []

@app.websocket("/detect")
async def websocket_endpoint(websocket: WebSocket):
    global plots
    try:
    ...
```
This tells python to treat it as globally scoped even though there is an assignment to the variable in the function body.

Also, you'll probably want to delete the `plots = []` line. (If you want to reset the list of plots, and run into trouble using `plots = []`, maybe try calling `plots.clear()` instead. This will remove items from the existing list rather than creating a new one. I'm not sure if this is necessary with your current design; haven't thought hard enough about it.)

-------

All this said, using global variables like this is definitely a "hack", and probably shouldn't be a part of a sustainable design.

But I think it's fine if you are just prototyping or trying things out! It's a hobby project after all.

**Comment by dmontagu** [comment]
Also, I feel compelled to add, you should *definitely **definitely \*never\**** be including this line in an endpoint anywhere that you don't 100% trust everything that can access the endpoint:

```
image = pickle.loads(request.get_data())
```

This is a common security mistake, and exposes arbitrary remote code execution to anyone that can access the endpoint.

Just in case you weren't aware!

(Again, it's more or less fine if you are not exposing the service in any way externally, and no one else is running it in an environment you don't control. But if anyone malicious were to see that line of code and be able to hit the endpoint they could basically have instant control of the machine - or at least the user running the service.)

**Comment by tiangolo** [comment]
Thanks for the help here everyone! :clap: :bow: 

Thanks for coming back and closing the issue @rlewkowicz  :+1:
